### PR TITLE
Example of MNIST using RNN

### PR DESCRIPTION
**Pull request by rakesh-malviya**
**Motivation:** Create pytorch example similar to Official Tensorflow Keras RNN example using MNIST [here](https://www.tensorflow.org/guide/keras/rnn)

I have written and tested the code by modifying following example on MNIST [here](https://github.com/pytorch/examples/tree/master/mnist) .

### CODE REVIEW

File: mnist_rnn/main.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                       ])),
+        batch_size=args.test_batch_size, shuffle=True, **kwargs)
+
+    # for i, (images, labels) in enumerate(train_loader):
+    #     print(images.shape)
+    #     exit()
+
+    model = Net().to(device)
+    optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
+
```
[the remark this answers is not in the document]
Comment (rakesh-malviya) [reply mnist_rnn/main.py:122]: I have removed excess lines

File: mnist_rnn/main.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+class Net(nn.Module):
+    def __init__(self):
+        super(Net, self).__init__()
+        self.rnn = nn.RNN(input_size=28,hidden_size=28, batch_first=True)
+        self.batchnorm = nn.BatchNorm1d(28)
+        self.dense = nn.Linear(28, 10)
+
+
+    def forward(self, x):
+        # shape of x is (batch_size,1,28,28)
```
[the remark this answers is not in the document]
Comment (rakesh-malviya) [reply mnist_rnn/main.py]: Intermediate variables are named 'output' in intermediate variables, I have made changes accordingly.

File: mnist_rnn/main.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        # converting shape of input to (batch_size, 28,28)
+        # as required by RNN when batch_first is set True
+        input = input.reshape(-1,28,28)
+        output, hidden = self.rnn(input)
+
+        # RNN output shape is (seq_len, batch, input_size)
+        # Get last output of RNN
+        output = output[:,-1,:]
+        output = self.batchnorm(output)
+        output = self.dropout1(output)
```
Comment (rakesh-malviya) [inline mnist_rnn/main.py:33]: Following changes made similar to pytorch MNIST example [here](https://github.com/pytorch/examples/blob/master/mnist/main.py#L27).

File: mnist_rnn/main.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.optim.lr_scheduler import StepLR
+
+
+class Net(nn.Module):
+    def __init__(self):
+        super(Net, self).__init__()
+        self.rnn = nn.LSTM(input_size=28,hidden_size=64, batch_first=True)
```
Comment (rakesh-malviya) [inline mnist_rnn/main.py:14]: Change RNN type from simple/basic RNN to LSTM since it gives better accuracy.

File: mnist_rnn/main.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                       ])),
+        batch_size=args.batch_size, shuffle=True, **kwargs)
+    test_loader = torch.utils.data.DataLoader(
+        datasets.MNIST('../data', train=False, transform=transforms.Compose([
+                           transforms.ToTensor(),
+                           transforms.Normalize((0.1307,), (0.3081,))
+                       ])),
+        batch_size=args.test_batch_size, shuffle=True, **kwargs)
+
+    # for i, (images, labels) in enumerate(train_loader):
```
[the remark this answers is not in the document]
Comment (rakesh-malviya) [reply mnist_rnn/main.py]: Removed commented code.

**Comment by rakesh-malviya** [comment]
Hi @osalpekar , I have resolved review comments.

Regards,
Rakesh

File: mnist_rnn/main.py (hunk 4 above)
Comment (mrshenli) [inline mnist_rnn/main.py:14]: nit: missing a space after the first comma

File: mnist_rnn/main.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        self.batchnorm = nn.BatchNorm1d(64)
+        self.dropout1 = nn.Dropout2d(0.25)
+        self.dropout2 = nn.Dropout2d(0.5)
+        self.fc1 = nn.Linear(64, 32)
+        self.fc2 = nn.Linear(32, 10)
+
+
+    def forward(self, input):
+        # Shape of input is (batch_size,1,28,28)
+        # converting shape of input to (batch_size, 28,28)
```
Comment (mrshenli) [inline mnist_rnn/main.py:24]: nit: shall we stay consistent in whether uses space or not. The line above does not use spaces, this line uses one, and a few lines below uses one space after every comma.

File: mnist_rnn/main.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                        help='learning rate (default: 0.1)')
+    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
+                        help='Learning rate step gamma (default: 0.7)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
+    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
+                        help='how many batches to wait before logging training status')
+
```
Comment (mrshenli) [inline mnist_rnn/main.py:95]: nit: unintentional new line?

File: mnist_rnn/main.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
+                        help='input batch size for training (default: 64)')
+    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
+                        help='input batch size for testing (default: 1000)')
+    parser.add_argument('--epochs', type=int, default=14, metavar='N',
+                        help='number of epochs to train (default: 14)')
+    parser.add_argument('--lr', type=float, default=0.1, metavar='LR',
+                        help='learning rate (default: 0.1)')
+    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
+                        help='Learning rate step gamma (default: 0.7)')
```
Comment (mrshenli) [inline mnist_rnn/main.py:88]: nit: let's be consistent on whether the help message should start from a capital letter.

Comment (rakesh-malviya) [reply mnist_rnn/main.py]: I have made help all message should start from a small letter.

**Comment by rakesh-malviya** [comment]
@zhangguanheng66 yes we can try `Transformer` in another example. 

Most people think MNIST is good introductory dataset for RNN and it is also one of the official RNN examples of `Keras tensorflow` as shown [here](https://www.tensorflow.org/guide/keras/rnn#load_mnist_dataset)

I am not sure if MNIST is good introductory dataset for `Transformer`

**Comment by rakesh-malviya** [comment]
> LGTM! Thanks for contributing! I added some cosmetic comments on formatting.
> 
> One general comment on the folder structure is whether we should move the three mnist examples from the root into a common parent folder.
> 
> And I am not sure who will be giving stamps to example contributions for vision applications.
> 
> cc @fmassa (for vision app) @zhangguanheng66 (for RNN) @soumith (general guidelines)
Hi @mrshenli ,

Thanks for reviews.

I suggest that root folder should correspond to specific Data science area or ML algorithm, since user intent to visit an example will be based on such topics.

Regards,
Rakesh

**Comment by rakesh-malviya** [comment]
> LGTM! Thanks for contributing! I added some cosmetic comments on formatting.
> 
> One general comment on the folder structure is whether we should move the three mnist examples from the root into a common parent folder.
> 
> And I am not sure who will be giving stamps to example contributions for vision applications.
> 
> cc @fmassa (for vision app) @zhangguanheng66 (for RNN) @soumith (general guidelines)

Hi @soumith , cc @mrshenli ,

We are not sure who will be giving stamps to this example contribution. This example uses `MNIST`(vision) dataset and it is good introductory example of `RNN`.

Thanks and regards,
Rakesh

**Comment by hudeven** [comment]
LGTM. Let's merge it!
